runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -1649,12 +1661,25 @@ obj:;
 			rc = GOTO_ASYNC_CHECK;
 			goto done;
 		}
-		if (0 == monitorRC) {
-			/* Monitor was not entered - hide the frame to prevent exception throw from processing it.
-			 * Note that BP can not have changed during a failed enter.
-			 */
+
+		if (monitorRC < J9_OBJECT_MONITOR_BLOCKING) {
 			*bp |= J9SF_A0_INVISIBLE_TAG;
-			rc = THROW_MONITOR_ALLOC_FAIL;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			if (J9_OBJECT_MONITOR_VALUE_TYPE_IMSE == monitorRC) {
+				J9UTF8 *badClassName = J9ROMCLASS_CLASSNAME(J9OBJECT_CLAZZ(_currentThread, syncObject)->romClass);
```

Comment:
syncObject may no longer be valid after enterObjectMonitor.

Reply:
Hmm, maybe this is OK based on the comment above about not releasing VM access in any failure case.